data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,675 @@
+
+A local definition may shadow any previous definition. Inner shadowing is limited to the local definition's scope. Therefore, anything written after will still take the previous definition, as shown:
+```ocaml
+# let d = 21;;
+val d : int = 21
+
+# let d = 7 in d * 2;;
+- : int = 14
+
+# d;;
+- : int = 21
+```
+
+Here, the value of `d` hasn't changed. It's still `21`, as defined in the first expression. The second expression binds `d` locally, inside `d * 2`, not globally.
+
+A name-value pair in a local expression *shadows* a binding with the same name in the global environment. In other words, the local binding temporarily hides the global one, making it inaccessible, but it doesn't change it.
+
+### Same-Level Shadowing
+
+Another kind of shadowing takes place when there are two definitions with the same name at the same level.
+```ocaml
+# let a = 2 * 3;;
+val a : int = 6
+
+# let e = a * 7;;
+val e : int = 42
+
+# let a = 7;;
+val a : int = 7
+
+# e;;
+- : int = 42
+
+# a;;
+- : int = 7
+```
+
+In this example, `a` is defined twice. The key thing to understand is that `a` is *not updated*. It looks as if the value of `a` has changed, but it hasn't. When the second `a` is defined, the first one becomes unreachable, but it remains in the global environment. This means anything written after the second definition uses its value, but functions written *before* the second definition still use the first, even if called later.
+
+## Function as Values
+
+As already stated, in OCaml, functions are values. This is the key concept of [functional programming](https://en.wikipedia.org/wiki/Functional_programming). In this context, it is also possible to say that OCaml has [first-class](https://en.wikipedia.org/wiki/First-class_function) functions or that functions are [first-class citizens](https://en.wikipedia.org/wiki/First-class_citizen).
+
+For now, let's put aside those definitions and instead start playing with functions. Their meaning will arise from experience. Once things make sense, using these terms is just a means to interact with the community.
+
+This is a big takeaway. We believe functional programming (_ergo_ OCaml) is best understood by practising it rather than reading about it. Just like skateboarding, cooking, or woodworking. Learning by doing is not only possible, it is usually the easiest and the best way to start. Because in the end, the goal is not to be able to give the correct definition of abstract terms but to write OCaml code.
+
+## Applying Functions
+
+When several expressions are written side by side, the leftmost one should be a function. All the others are parameters. In OCaml, no parentheses are needed to express passing an argument to a function. Parentheses serve a single purpose: associating expressions to create subexpressions.
+```ocaml
+# max (21 * 2) (int_of_string "713");;
+- : int = 713
```

Comment:
1. Yes `max` is already defined
2. Evaluation has 3 steps here
  1. `int_of_string "713"` : the string "713" is turned into the integer 713
  2. `21 * 2` : this product is evaluated, result value is 42
  3. `max 42 713` is computed, since 713 > 42, the final result is 713

Reply:
Yes `max` is already available, it is predefined

Here is how evaluation takes place:

1. `int_of_string "713"` results in 713 (string converted to integer)
2. `21 * 2` results in 42 (multiplication computed)
3. `max 42 713` results in 713 (comparison of 42 and 713, the latter is the bigger)